core/trino-main/src/test/java/io/trino/sql/query/TestListagg.java

```diff
@@ -0,0 +1,243 @@
+                "                   (1, 1, 'a'), " +
+                "                   (1, 3, 'c'), " +
+                "                   (1, 2, 'b'), " +
+                "                   (2, 6, 'f'), " +
+                "                   (2, 5, 'e'), " +
+                "                   (2, 4, 'd') " +
+                "               ) t(id, o, value)" +
+                "          ORDER BY id, o"))
+                .matches("VALUES     " +
+                        "     (1, 1, CAST('a' as varchar))," +
+                        "     (1, 2, CAST('a,b' as varchar))," +
+                        "     (1, 3, CAST('a,b,c' as varchar))," +
+                        "     (2, 4, CAST('d' as varchar))," +
+                        "     (2, 5, CAST('d,e' as varchar))," +
+                        "     (2, 6, CAST('d,e,f' as varchar))");
+    }
+
+    @Test
+    public void testListaggQueryOverflowError()
+    {
+        String tooLargeValue = StringUtils.repeat("a", DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+        assertThatThrownBy(() -> assertions.query("SELECT LISTAGG(value, ',' ON OVERFLOW ERROR)" +
+                "          FROM (VALUES " +
+                "                  ('" + tooLargeValue + "')," +
+                "                  ('Trino')" +
+                "               ) t(value) "))
+                .isInstanceOf(TrinoException.class)
+                .hasMessage("Concatenated string is too large")
+                .matches(throwable -> ((TrinoException) throwable).getErrorCode() == QUERY_REJECTED.toErrorCode());
+    }
+
+    @Test
+    public void testListaggQueryOverflowTruncateWithoutCountAndWithoutTruncationFiller()
+    {
+        String largeValue = StringUtils.repeat("a", DEFAULT_MAX_PAGE_SIZE_IN_BYTES - 6);
+        assertThat(assertions.query("SELECT LISTAGG(value, ',' ON OVERFLOW TRUNCATE WITHOUT COUNT)" +
+                "          FROM (VALUES " +
+                "                  ('" + largeValue + "')," +
+                "                  ('Trino')," +
+                "                  ('rocks')" +
+                "               ) t(value) "))
+                .matches("VALUES     " +
+                        "     (CAST('" + largeValue + ",Tr...' as varchar))");
+    }
+
+    @Test
+    public void testListaggQueryOverflowTruncateWithCountAndWithTruncationFiller()
+    {
+        String largeValue = StringUtils.repeat("a", DEFAULT_MAX_PAGE_SIZE_IN_BYTES - 12);
+        assertThat(assertions.query("SELECT LISTAGG(value, ',' ON OVERFLOW TRUNCATE '.....' WITH COUNT)" +
+                "          FROM (VALUES " +
+                "                  ('" + largeValue + "')," +
+                "                  ('Trino')," +
+                "                  ('SQL')," +
+                "                  ('Engine')" +
+                "               ) t(value) "))
+                .matches("VALUES     " +
+                        "     (CAST('" + largeValue + ",Tri.....[2]' as varchar))");
```

Comment:
Per the SQL spec, the count should be enclosed in parenthesis:

<img width="721" alt="Screen Shot 2021-05-25 at 3 28 57 PM" src="https://user-images.githubusercontent.com/14387/119576638-f6861780-bd6d-11eb-8451-5de6c0a98eb7.png">

Reply:
Can you share a link where to download the specifications?

Reply:
Unfortunately it’s not freely available. You need to purchase it from the ISO or ANSI organizations:

https://www.iso.org/obp/ui/#iso:std:iso-iec:9075:-2:ed-5:v1:en